descartes_trajectory/src/cartesian_interpolator.cpp

```diff
@@ -0,0 +1,199 @@
+/*
+ * Software License Agreement (Apache License)
+ *
+ * Copyright (c) 2014, Southwest Research Institute
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <console_bridge/console.h>
+#include <descartes_trajectory/cartesian_interpolator.h>
+#include <tf_conversions/tf_eigen.h>
+#include <boost/uuid/uuid_io.hpp>
+#include <math.h>
+
+
+#define NOT_IMPLEMENTED_ERR(ret) logError("%s not implemented", __PRETTY_FUNCTION__); return ret;
+const static unsigned int MIN_NUM_STEPS = 1;
+
+using namespace descartes_core;
+
+namespace descartes_trajectory
+{
+LinearSegment::LinearSegment(const tf::Transform& start_pose,
+                             const tf::Transform& end_pose,
+                             double total_time,
+                             double interpolation_step):
+                             start_pose_(start_pose),
+                             end_pose_(end_pose),
+                             total_time_(total_time),
+                             interpolation_step_(interpolation_step)
+{
+
+}
+
+LinearSegment::~LinearSegment()
+{
+
+}
+
+bool LinearSegment::interpolate(std::vector<descartes_core::TrajectoryPtPtr>& segment_points)
+{
+
+  tf::Vector3 vel = (end_pose_.getOrigin() - start_pose_.getOrigin())/total_time_;
+  double time_step = interpolation_step_/total_time_;
+  std::size_t num_steps = std::ceil(total_time_/interpolation_step_)-1;
+
+  if(num_steps <= MIN_NUM_STEPS)
+  {
+    ROS_ERROR_STREAM("Interpolation steps value of "<<num_steps<<" is invalid");
+    return false;
+  }
+  segment_points.clear();
+
+  segment_points.reserve(num_steps);
+  tf::Transform p;
+  Eigen::Affine3d eigen_pose;
+  tf::Quaternion q;
+  for(std::size_t i = 1; i < num_steps; i++)
+  {
+    p.setOrigin(start_pose_.getOrigin() +  vel*i*interpolation_step_);
+    start_pose_.getRotation().slerp(q,i*time_step);
+    tf::poseTFToEigen(p,eigen_pose);
+    segment_points.push_back(TrajectoryPtPtr(new CartTrajectoryPt(eigen_pose)));
+  }
+
+  ROS_DEBUG_STREAM("Segment contains "<<segment_points.size()<<" interpolated points for "<<num_steps<<" time steps");
+
+  return true;
+}
+
+CartesianInterpolator::CartesianInterpolator()
```

Comment:
Is this constructor hidden?  Using this constructor will result in undefined class variables (or rather random values).